compiler-core/src/language_server/code_action.rs

```diff
@@ -2852,6 +2852,277 @@ impl<'ast> ast::visit::Visit<'ast> for ExtractVariable<'ast> {
+            .changes(self.params.text_document.uri.clone(), self.edits.edits)
+            .preferred(false)
+            .push_to(&mut action);
+        action
+    }
+}
+
+impl<'ast> ast::visit::Visit<'ast> for ExtractConstant<'ast> {
+    /// To get the position of the function containing the value or assignment
+    /// to extract
+    fn visit_typed_function(&mut self, fun: &'ast ast::TypedFunction) {
+        let fun_location = fun.location;
+        let fun_range = self.edits.src_span_to_lsp_range(SrcSpan {
+            start: fun_location.start,
+            end: fun.end_position,
+        });
+
+        if !within(self.params.range, fun_range) {
+            return;
+        }
+        self.container_function_start = Some(fun.location.start);
+
+        ast::visit::visit_typed_function(self, fun);
+    }
+
+    /// To extract the whole assignment
+    fn visit_typed_assignment(&mut self, assignment: &'ast TypedAssignment) {
+        let expr_location = assignment.location;
+
+        // We only offer this code action for extracting the whole assignment
+        // between `let` and `=`.
+        let pattern_location = assignment.pattern.location();
+        let location = SrcSpan::new(assignment.location.start, pattern_location.end);
+        let code_action_range = self.edits.src_span_to_lsp_range(location);
+
+        if !within(self.params.range, code_action_range) {
+            ast::visit::visit_typed_assignment(self, assignment);
+            return;
+        }
+
+        if assignment.pattern.is_variable()
+            && is_literal_or_made_of_literals(assignment.value.as_ref())
+        {
+            let selected_expression = EcoString::from(
+                self.module
+                    .code
+                    .get(
+                        (assignment.value.location().start as usize)
+                            ..(assignment.location.end as usize),
+                    )
+                    .expect("selected expression"),
+            );
+
+            self.type_of_extractable = Some(ExtractableToConstant::Assignment);
+            self.selected_expression = Some((expr_location, assignment.type_()));
+            self.name_to_use = match &assignment.pattern {
+                Pattern::Variable { name, .. } => Some(name.clone()),
+                _ => None,
+            };
+            self.value_to_use = Some(selected_expression);
+        }
+    }
+
+    /// To extract only the literal
+    fn visit_typed_expr(&mut self, expr: &'ast TypedExpr) {
+        let expr_location = expr.location();
+        let expr_range = self.edits.src_span_to_lsp_range(expr_location);
+
+        if !within(self.params.range, expr_range) {
+            ast::visit::visit_typed_expr(self, expr);
+            return;
+        }
+
+        if matches!(
+            self.type_of_extractable,
+            None | Some(ExtractableToConstant::Collection)
+        ) && is_literal_or_made_of_literals(expr)
```

Comment:
What is this bit doing? Would be useful to have documentation comments explaining the intent